Pelias/custom address fields

mrbig opened the pull request:
Pelias is an open source modular geocoder. It support several data sources and tries to find the best match for each query. Unfortunately a lot of this information is hidden in PHP Geocoder. This PR will expose the most important fields through an extended PeliasAddress class.

The solution is similar to the GoogleMaps provider's GoogleAddress, I have also included the with... methods although currently they are not in use.

The change also includes a minor refactor in the provider that leads to better extendability if one would want to override how the results are parsed into addresses.

norkunas commented on src/Provider/Pelias/Model/PeliasAddress.php:
```diff
@@ -0,0 +1,175 @@
+{
+    /**
+     * The pelias layer returned
+     * @var string|null
+     */
+    private $layer;
```
Why not defining real typehints?

mrbig replied:
I have added real type hints.

norkunas commented on src/Provider/Pelias/Pelias.php:
```diff
@@ -132,48 +133,62 @@ protected function executeQuery(string $url): AddressCollection
+     */
+    protected function buildAddress(array $location)
+    {
+        if (isset($location['bbox'])) {
+            $bounds = [
+                'south' => $location['bbox'][3],
+                'west' => $location['bbox'][2],
+                'north' => $location['bbox'][1],
+                'east' => $location['bbox'][0],
+            ];
+        } else {
+            $bounds = [
+                'south' => null,
+                'west' => null,
+                'north' => null,
+                'east' => null,
+            ];
+        }
```
```suggestion
        $bounds = [
            'south' => $location['bbox'][3] ?? null,
            'west' => $location['bbox'][2] ?? null,
            'north' => $location['bbox'][1] ?? null,
            'east' => $location['bbox'][0] ?? null,
        ];
```

mrbig replied:
My goal was to change the original Provider as minimally as possible.
However this is a nice touch to make the code more readable, so I have included your suggestion.

norkunas commented on src/Provider/Pelias/Pelias.php:
```diff
@@ -132,48 +133,62 @@ protected function executeQuery(string $url): AddressCollection
+            'latitude' => $location['geometry']['coordinates'][1],
+            'longitude' => $location['geometry']['coordinates'][0],
+            'bounds' => $bounds,
+            'streetNumber' => isset($props['housenumber']) ? $props['housenumber'] : null,
```
null coalescing could be used

mrbig replied:
I have replaced the code wherever possible

jbelien reviewed (changes requested):
Thanks @mrbig !

Could you add documentation for the properties you've added. Are those properties **always** return by any Pelias instance ?

Please also have a look at the comments made by @norkunas.

Thanks!

mrbig commented:
@jbelien I have implemented the changes suggested by @norkunas.

Could you point me to the documentation that I should update? I've included documentation in the code itself, I realize that might not be sufficient.

The fields I have added are always returned to my knowledge. They describe what data source was used, how accurate that data source is considered, could it find the exact location or it did some kind of interpolation.

Our use case:
We are geocoding a large dataset using opensource data. Unfortunately some regions aren't exact in these databases at the moment. So we want to store with the coordinates the accuracy at the time of geocoding, so that after a map update we can go reiterate these addresses and run another geocoding to refine them. But if the result of geocoding change without the accuracy information we even can not decide if the new values are better or worst.

jbelien commented:
> Could you point me to the documentation that I should update? I've included documentation in the code itself, I realize that might not be sufficient.

In `src/Provider/Pelias/Model/PeliasAddress.php`, could you add the link to the documentation ?

For instance:
```php
/**
 * @link https://github.com/pelias/documentation/blob/master/result_quality.md#confidence Confidence score
 * @var float|null
 */
private ?float $confidence = null;
```

mrbig commented:
Hi @jbelien do you need any other change on this PR?
